trace/tracestate.go

```diff
@@ -39,43 +33,115 @@ const (
+func checkValueLast(v byte) bool {
+	return v >= '\x21' && v <= '\x7e' && v != '\x2c' && v != '\x3d'
+}
+
+// based on the W3C Trace Context specification, see
+// https://www.w3.org/TR/trace-context-1/#value
+// value    = (0*255(chr)) nblk-chr
+// nblk-chr = %x21-2B / %x2D-3C / %x3E-7E
+// chr      = %x20 / nblk-chr
+func checkValue(val string) bool {
+	n := len(val)
+	if n == 0 || n > 256 {
+		return false
 	}
-	if !noTenantKeyRe.MatchString(key) {
-		if !withTenantKeyRe.MatchString(key) {
-			return member{}, fmt.Errorf("%w: %s", errInvalidKey, key)
+	for i := 0; i < n-1; i++ {
+		if !checkValueChar(val[i]) {
+			return false
 		}
-		atIndex := strings.LastIndex(key, "@")
-		if atIndex > 241 || len(key)-1-atIndex > 14 {
-			return member{}, fmt.Errorf("%w: %s", errInvalidKey, key)
+	}
+	return checkValueLast(val[n-1])
+}
+
+func checkKeyRemain(key string) bool {
+	// ( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )
+	for _, v := range key {
+		if (v >= '0' && v <= '9') || (v >= 'a' && v <= 'z') {
+			continue
+		}
+		switch v {
+		case '_', '-', '*', '/':
+			continue
 		}
+		return false
+	}
+	return true
+}
+
+// simple-key = lcalpha (0*255( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ))
+// system-id = lcalpha (0*13( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ))
+// n is remain part length, should be 255 in simple-key or 13 in system-id
+func checkKeyPart(key string, n int) bool {
+	if len(key) == 0 {
+		return false
+	}
+	first := key[0] // key's first char
+	ret := len(key[1:]) <= n
+	ret = ret && first >= 'a' && first <= 'z'
+	return ret && checkKeyRemain(key[1:])
+}
+
+// tenant-id = ( lcalpha / DIGIT ) 0*240( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )
+// n is remain part length, should be 240 exactly
+func checkKeyTenant(key string, n int) bool {
+	if len(key) == 0 {
+		return false
+	}
+	first := key[0] // key's first char
+	ret := len(key[1:]) <= n
+	ret = ret && ((first >= 'a' && first <= 'z') || (first >= '0' && first <= '9'))
+	return ret && checkKeyRemain(key[1:])
+}
+
+// based on the W3C Trace Context specification, see
+// https://www.w3.org/TR/trace-context-1/#tracestate-header
+//
+// key = simple-key / multi-tenant-key
+// simple-key = lcalpha (0*255( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ))
+// multi-tenant-key = tenant-id "@" system-id
+// tenant-id = ( lcalpha / DIGIT ) (0*240( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ))
+// system-id = lcalpha (0*13( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ))
+// lcalpha    = %x61-7A ; a-z
+func checkKey(key string) bool {
+	tenant, system, ok := strings.Cut(key, "@")
+	if !ok {
+		return checkKeyPart(key, 255)
 	}
-	if len(value) > 256 || !valueRe.MatchString(value) {
+	return checkKeyTenant(tenant, 240) && checkKeyPart(system, 13)
+}
+
+func newMember(key, value string) (member, error) {
+	if !checkKey(key) {
+		return member{}, fmt.Errorf("%w: %s", errInvalidKey, key)
+	}
+	if !checkValue(value) {
 		return member{}, fmt.Errorf("%w: %s", errInvalidValue, value)
```

Comment:
It seems like 2 allocations are coming from each of these returned errors constructed with `fmt.Errorf`. Both `key` and `value` are escaping to the heap in the call and a string is being allocated (AFAIK). Though, the returned error here is not used by `parseMember`.

It would be worth looking into just returning either `errInvalidKey` or `errInvalidValue` here and constructing an appropriate error in `parseMember` given you know key and value there.